rust/main/chains/hyperlane-ethereum/src/signer/yubi.rs

```diff
@@ -0,0 +1,146 @@
+use ::ecdsa::elliptic_curve::consts::U32;
+use ethers_core::{k256::PublicKey, types::Address, utils::keccak256};
+use ethers_signers::Wallet;
+use signature::digest::Digest;
+use signature::digest::FixedOutput;
+use signature::{DigestSigner, Error};
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+use std::sync::{LazyLock, Mutex};
+use yubihsm::ecdsa::sec1::FromEncodedPoint;
+use yubihsm::ecdsa::sec1::ToEncodedPoint;
+use yubihsm::object;
+use yubihsm::Client;
+use yubihsm::HttpConfig;
+use yubihsm::{ecdsa::Signer as YubiSigner, Credentials};
+
+/// Unique key for a yubihsm client in the global cache.
+/// The global caches is used so that we can reuse yubihsm sessions across threads, preventing resource exhaustion.
+struct YubihsmClientCacheKey {
+    config: HttpConfig,
+    credentials: Credentials,
+    id: object::Id,
+}
+
+impl PartialEq for YubihsmClientCacheKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.config.addr == other.config.addr
+            && self.config.port == other.config.port
+            && self.config.timeout_ms == other.config.timeout_ms
+            && self.credentials.authentication_key_id == other.credentials.authentication_key_id
+            && self.id == other.id
+    }
+}
+
+impl Eq for YubihsmClientCacheKey {}
+
+impl Hash for YubihsmClientCacheKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.config.addr.hash(state);
+        self.config.port.hash(state);
+        self.config.timeout_ms.hash(state);
+        self.credentials.authentication_key_id.hash(state);
+        self.id.hash(state);
+    }
+}
+
+/// A thread safe cache for signers against a yubihsm.
+/// Yubihsms have a limited number of sessions (16), which remain open for at least 30 seconds. If a number of connections is requested
+/// too quickly, the devices resources will become exhausted leading to failures.
+static YUBIHSM_CLIENT_CACHE: LazyLock<Mutex<HashMap<YubihsmClientCacheKey, WrappedSigner>>> =
```

Comment:
[Original Comment](https://github.com/Tessellated-io/hyperlane-monorepo/pull/2#discussion_r1979728599)

> I wonder if it would make sense to get a shared access for readers and exclusive access to writers. I suppose that majority of the times we need only read a value from this hash map. Probably, RwLock is more appropriate here?

Done